File: backend/app/domain/flex_fit.py
"""按内容自然高度回填 column grow，避免内容少的块白占大片空间。

solver 只按权重瓜分画布、不读字体度量，所以预设里写死的 grow 会让
「3 条要点」和「8 条要点」占一样高。这里在生成期用 text_metrics 估算每个
叶子的自然高度，换算成 grow；富余高度交给可吸收的块（图片/图表）或尾部
占位块，而不是摊回文字块。

纯函数，只改 grow 与尾部占位块，不动块内容与 row 的横向比例。
"""

from __future__ import annotations

from app.domain.block_style import content_rect_pt, merge_text_style, resolve_box
from app.domain.content import Block
from app.domain.flex_layout import FlexContainer, FlexLeaf, FlexNode, is_spacer
from app.domain.flex_normalize import GROW_MAX, GROW_MIN, TITLE_STYLES, normalize
from app.domain.flex_solve import solve
from app.domain.geometry import (
    CANVAS_HEIGHT_PT,
    CANVAS_WIDTH_PT,
    SAFE_AREA_HEIGHT_PT,
    SAFE_AREA_WIDTH_PT,
)
from app.domain.text_metrics import TEXTBOX_MARGIN_PT, measure_bullets, measure_text
from app.domain.theme import Theme

# 文字块在自然高度上留一点呼吸余量，避免贴着字挤成一条
_BREATHING = 1.12

# 没有可度量内容的块给一个偏好高度（pt，相对 540 高画布）
_PREFERRED_HEIGHT_PT: dict[str, float] = {
    "image": 224.0,
    "chart": 240.0,
    "kpi": 96.0,
    "callout": 48.0,
}
_TABLE_ROW_HEIGHT_PT = 26.0
_CARD_BASE_HEIGHT_PT = 88.0
_CARD_LINE_HEIGHT_PT = 22.0
_MIN_LEAF_HEIGHT_PT = 36.0

# 这些块可以无限吃掉富余高度：拉大反而更好看
_ABSORBING_TYPES = frozenset({"image", "chart"})

# 有限吸收：可在偏好高度的这个倍数内吃富余（KPI/表/卡片拉高观感更好）
_LIMITED_ABSORB_TYPES = frozenset({"kpi", "table", "cards"})
_LIMITED_ABSORB_MAX_RATIO = 1.6

# 富余高度小于列高的这个比例时不再造占位块，直接摊给内容
_SURPLUS_MIN_RATIO = 0.06

# 尾部占位块最多吃掉列高的这个比例：只做收边，不吞下半页
_MAX_SPACER_RATIO = 0.16

# 封面/章节页不铺满，而是把富余拆到内容上下两侧；上方占这个比例，
# 让标题落在视觉中线偏上，与固定布局 cover.json 的观感一致
_CENTERED_ROLES = frozenset({"cover", "section"})
_LEAD_SPACER_RATIO = 0.38


def fit_tree_to_content(
    tree: FlexContainer,
    blocks: list[Block],
    *,
    theme: Theme,
    page_role: str | None = None,
) -> FlexContainer:
    """返回 grow 已按内容自然高度重算的新树。

    只在生成期调用：用户手动调过版面的页面不应被重新压扁。
    """
    block_map = {block.id: block for block in blocks}
    # 叶子宽度只由 row ratios 决定，与 grow 无关，因此求解一次即可定
    widths_pt = {
        placed.block_id: placed.rect.w * CANVAS_WIDTH_PT for placed in solve(tree)
    }
    ctx = _FitContext(
        block_map=block_map,
        widths_pt=widths_pt,
        theme=theme,
        centered=page_role in _CENTERED_ROLES,
    )
    fitted = _fit_container(tree, SAFE_AREA_HEIGHT_PT, ctx, is_root=True)
    # 高度来自真实度量，不需要再按标题上限回钳
    return normalize(fitted, clamp_title_grow=False)


class _FitContext:
    def __init__(
        self,
        *,
        block_map: dict[str, Block],
        widths_pt: dict[str, float],
        theme: Theme,
        centered: bool = False,
    ) -> None:
        self.block_map = block_map
        self.widths_pt = widths_pt
        self.theme = theme
        self.centered = centered


def _fit_container(
    node: FlexContainer,
    avail_h_pt: float,
    ctx: _FitContext,
    *,
    is_root: bool = False,
) -> FlexContainer:
    if not node.children:
        return node

    if node.type == "row":
        # 同一行的子项共享整行高度，横向比例不在本函数职责内
        return node.model_copy(
            update={
                "children": [_fit_child(child, avail_h_pt, ctx) for child in node.children]
            }
        )

    children = list(node.children)
    gaps_pt = max(len(children) - 1, 0) * node.gap_pt
    targets = [_natural_height_pt(child, ctx) for child in children]
    surplus = avail_h_pt - gaps_pt - sum(targets)

    if surplus > 0 and is_root and ctx.centered:
        children, targets = _center_vertically(children, targets, surplus, node.gap_pt)
    elif surplus > 0:
        children, targets = _absorb_surplus(
            children, targets, surplus, avail_h_pt, node.gap_pt, ctx
        )

    grows = _targets_to_grows(targets)
    fitted = [
        _fit_child(child, target, ctx).model_copy(update={"grow": grow})
        for child, target, grow in zip(children, targets, grows, strict=True)
    ]
    return node.model_copy(update={"children": fitted})


def _fit_child(child: FlexNode, avail_h_pt: float, ctx: _FitContext) -> FlexNode:
    if isinstance(child, FlexContainer):
        return _fit_container(child, avail_h_pt, ctx)
    return child


def _absorb_surplus(
    children: list[FlexNode],
    targets: list[float],
    surplus: float,
    avail_h_pt: float,
    gap_pt: float,
    ctx: _FitContext,
) -> tuple[list[FlexNode], list[float]]:
    """把富余高度交给吸收型块或尾部占位块，不摊回文字块。"""
    remaining = surplus

    # 1) 图片/图表无限吸收
    absorbers = [index for index, child in enumerate(children) if _is_absorbing(child, ctx)]
    if absorbers:
        return children, _distribute(targets, remaining, absorbers)

    # 2) KPI/表/卡片有限吸收（偏好高度 × 1.6）
    limited = [
        index for index, child in enumerate(children) if _is_limited_absorber(child, ctx)
    ]
    if limited:
        room = [
            max(0.0, targets[index] * _LIMITED_ABSORB_MAX_RATIO - targets[index])
            for index in limited
        ]
        take = min(remaining, sum(room))
        if take > 0:
            targets = _distribute_capped(targets, take, limited, room)
            remaining -= take

    if remaining <= 0:
        return children, targets

    # 3) 加一个占位块要多算一道间隙；富余不足时不值得造，直接摊给正文
    surplus_after_gap = remaining - gap_pt
    if surplus_after_gap < avail_h_pt * _SURPLUS_MIN_RATIO:
        body = [index for index, child in enumerate(children) if not _is_title(child)]
        return children, _distribute(targets, remaining, body or range(len(targets)))

    # 占位块只做收边；剩下的给正文而不是标题
    spacer_h = min(surplus_after_gap, avail_h_pt * _MAX_SPACER_RATIO)
    leftover = surplus_after_gap - spacer_h
    if leftover > 0:
        body = [index for index, child in enumerate(children) if not _is_title(child)]
        targets = _distribute(targets, leftover, body or range(len(targets)))

    return [*children, _fit_spacer("spacer-fit")], [*targets, spacer_h]


def _center_vertically(
    children: list[FlexNode],
    targets: list[float],
    surplus: float,
    gap_pt: float,
) -> tuple[list[FlexNode], list[float]]:
    """封面/章节页：富余拆到内容上下两侧，内容整体落在视觉中线偏上。"""
    budget = surplus - 2 * gap_pt
    if budget <= 0:
        return children, targets
    lead = budget * _LEAD_SPACER_RATIO
    return (
        [_fit_spacer("spacer-fit-lead"), *children, _fit_spacer("spacer-fit-trail")],
        [lead, *targets, budget - lead],
    )


def _fit_spacer(node_id: str) -> FlexContainer:
    return FlexContainer(type="column", id=node_id, children=[], gap_pt=0.0, grow=1.0)


def _distribute(targets: list[float], amount: float, indices) -> list[float]:
    """按现有目标高度的比例把 amount 分给指定下标。"""
    picked = list(indices)
    if not picked or amount <= 0:
        return list(targets)
    weight_sum = sum(targets[index] for index in picked)
    result = list(targets)
    if weight_sum <= 0:
        share = amount / len(picked)
        for index in picked:
            result[index] += share
        return result
    for index in picked:
        result[index] += amount * targets[index] / weight_sum
    return result


def _distribute_capped(
    targets: list[float],
    amount: float,
    indices: list[int],
    rooms: list[float],
) -> list[float]:
    """按各块剩余容量比例分配，且不超过各自 room。"""
    if not indices or amount <= 0:
        return list(targets)
    room_sum = sum(rooms)
    if room_sum <= 0:
        return list(targets)
    result = list(targets)
    for index, room in zip(indices, rooms, strict=True):
        result[index] += amount * room / room_sum
    return result


def _is_title(node: FlexNode) -> bool:
    return isinstance(node, FlexLeaf) and node.text_style in TITLE_STYLES


def _targets_to_grows(targets: list[float]) -> list[float]:
    """把目标高度换算成均值为 1 的 grow，并夹在 normalize 允许的区间内。"""
    n = len(targets)
    total = sum(targets)
    if n == 0:
        return []
    if total <= 0:
        return [1.0] * n
    return [min(GROW_MAX, max(GROW_MIN, target / total * n)) for target in targets]


def _is_absorbing(node: FlexNode, ctx: _FitContext) -> bool:
    """拉大反而更好看的节点：图片/图表，以及已有的占位块。"""
    if isinstance(node, FlexLeaf):
        block = ctx.block_map.get(node.block_id)
        return block is not None and block.type in _ABSORBING_TYPES
    if is_spacer(node):
        return True
    return any(_is_absorbing(child, ctx) for child in node.children)


def _is_limited_absorber(node: FlexNode, ctx: _FitContext) -> bool:
    if isinstance(node, FlexLeaf):
        block = ctx.block_map.get(node.block_id)
        return block is not None and block.type in _LIMITED_ABSORB_TYPES
    if is_spacer(node):
        return False
    return any(_is_limited_absorber(child, ctx) for child in node.children)


def _natural_height_pt(node: FlexNode, ctx: _FitContext) -> float:
    if isinstance(node, FlexLeaf):
        return _leaf_height_pt(node, ctx)

    if is_spacer(node) and not node.children:
        return 0.0
    if not node.children:
        return _MIN_LEAF_HEIGHT_PT

    heights = [_natural_height_pt(child, ctx) for child in node.children]
    if node.type == "row":
        return max(heights)
    return sum(heights) + max(len(heights) - 1, 0) * node.gap_pt


def _leaf_height_pt(leaf: FlexLeaf, ctx: _FitContext) -> float:
    block = ctx.block_map.get(leaf.block_id)
    if block is None:
        return _MIN_LEAF_HEIGHT_PT

    if block.type == "text":
        return _measured_height_pt(leaf, block, ctx, kind="text")
    if block.type == "bullets":
        return _measured_height_pt(leaf, block, ctx, kind="bullets")
    if block.type == "table":
        rows = 1 + len(block.rows)
        return max(_MIN_LEAF_HEIGHT_PT, rows * _TABLE_ROW_HEIGHT_PT)
    if block.type == "cards":
        # 横排卡片高度取最高一张的标题+描述估算
        tallest = max(
            (
                _CARD_BASE_HEIGHT_PT
                + max(0, (len(item.title) + len(item.desc)) // 28) * _CARD_LINE_HEIGHT_PT
            )
            for item in block.items
        )
        return max(_MIN_LEAF_HEIGHT_PT, tallest)
    if block.type == "callout":
        return _PREFERRED_HEIGHT_PT["callout"]
    return _PREFERRED_HEIGHT_PT.get(block.type, _MIN_LEAF_HEIGHT_PT)


def _measured_height_pt(
    leaf: FlexLeaf,
    block: Block,
    ctx: _FitContext,
    *,
    kind: str,
) -> float:
    width_pt = ctx.widths_pt.get(leaf.block_id, SAFE_AREA_WIDTH_PT)
    box = resolve_box(ctx.theme, block.style)
    # 度量只关心宽度；高度给足，避免把 overflow 判定卷进来
    avail_w, _ = content_rect_pt(width_pt, CANVAS_HEIGHT_PT, padding_pt=box.padding_pt)
    default_style = "bullet" if kind == "bullets" else "body"
    style = merge_text_style(ctx.theme, leaf.text_style or default_style, block.style)

    if kind == "bullets":
        used = measure_bullets(
            block.items, style=style, width_pt=avail_w, height_pt=CANVAS_HEIGHT_PT
        ).height_pt
    else:
        used = measure_text(
            block.text, style=style, width_pt=avail_w, height_pt=CANVAS_HEIGHT_PT
        ).height_pt

    chrome = 2 * TEXTBOX_MARGIN_PT + 2 * box.padding_pt
    return max(_MIN_LEAF_HEIGHT_PT, used * _BREATHING + chrome)

File: java_backend/src/main/java/com/aippt/domain/flex/FlexFit.java
package com.aippt.domain.flex;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

import com.aippt.domain.Geometry;
import com.aippt.domain.TextMetrics;
import com.aippt.domain.Theme;
import com.aippt.domain.content.BlockStyle;
import com.aippt.domain.content.Blocks;

public final class FlexFit {

    private static final double BREATHING = 1.12;
    private static final Map<String, Double> PREFERRED_HEIGHT_PT = Map.of(
            "image", 224.0,
            "chart", 240.0,
            "kpi", 96.0,
            "callout", 48.0
    );
    private static final double TABLE_ROW_HEIGHT_PT = 26.0;
    private static final double CARD_BASE_HEIGHT_PT = 88.0;
    private static final double CARD_LINE_HEIGHT_PT = 22.0;
    private static final double MIN_LEAF_HEIGHT_PT = 36.0;
    private static final Set<String> ABSORBING_TYPES = Set.of("image", "chart");
    private static final Set<String> LIMITED_ABSORB_TYPES = Set.of("kpi", "table", "cards");
    private static final double LIMITED_ABSORB_MAX_RATIO = 1.6;
    private static final double SURPLUS_MIN_RATIO = 0.06;
    private static final double MAX_SPACER_RATIO = 0.16;
    private static final Set<String> CENTERED_ROLES = Set.of("cover", "section");
    private static final double LEAD_SPACER_RATIO = 0.38;

    private FlexFit() {
    }

    public static FlexContainer fitTreeToContent(
            FlexContainer tree,
            List<Map<String, Object>> blocks,
            Theme theme,
            String pageRole
    ) {
        Map<String, Map<String, Object>> blockMap = new HashMap<>();
        for (Map<String, Object> block : blocks) {
            blockMap.put(Blocks.id(block), block);
        }
        Map<String, Double> widthsPt = new HashMap<>();
        for (FlexSolve.PlacedBlock placed : FlexSolve.solve(tree)) {
            widthsPt.put(placed.blockId(), placed.rect().w() * Geometry.CANVAS_WIDTH_PT);
        }
        FitContext ctx = new FitContext(blockMap, widthsPt, theme, CENTERED_ROLES.contains(pageRole));
        FlexContainer fitted = fitContainer(tree, Geometry.SAFE_AREA_HEIGHT_PT, ctx, true);
        return FlexNormalize.normalize(fitted, false);
    }

    private record FitContext(
            Map<String, Map<String, Object>> blockMap,
            Map<String, Double> widthsPt,
            Theme theme,
            boolean centered
    ) {
    }

    private static FlexContainer fitContainer(FlexContainer node, double availHPt, FitContext ctx, boolean isRoot) {
        if (node.children().isEmpty()) {
            return node;
        }
        if (node.isRow()) {
            List<FlexNode> children = new ArrayList<>();
            for (FlexNode child : node.children()) {
                children.add(fitChild(child, availHPt, ctx));
            }
            return node.withChildren(children);
        }
        List<FlexNode> children = new ArrayList<>(node.children());
        double gapsPt = Math.max(children.size() - 1, 0) * node.gapPt();
        List<Double> targets = new ArrayList<>();
        for (FlexNode child : children) {
            targets.add(naturalHeightPt(child, ctx));
        }
        double surplus = availHPt - gapsPt - targets.stream().mapToDouble(Double::doubleValue).sum();
        if (surplus > 0 && isRoot && ctx.centered()) {
            var centered = centerVertically(children, targets, surplus, node.gapPt());
            children = centered.children();
            targets = centered.targets();
        } else if (surplus > 0) {
            var absorbed = absorbSurplus(children, targets, surplus, availHPt, node.gapPt(), ctx);
            children = absorbed.children();
            targets = absorbed.targets();
        }
        List<Double> grows = targetsToGrows(targets);
        List<FlexNode> fitted = new ArrayList<>();
        for (int i = 0; i < children.size(); i++) {
            fitted.add(fitChild(children.get(i), targets.get(i), ctx).withGrow(grows.get(i)));
        }
        return node.withChildren(fitted);
    }

    private static FlexNode fitChild(FlexNode child, double availHPt, FitContext ctx) {
        if (child instanceof FlexContainer container) {
            return fitContainer(container, availHPt, ctx, false);
        }
        return child;
    }

    private record Pair(List<FlexNode> children, List<Double> targets) {
    }

    private static Pair absorbSurplus(
            List<FlexNode> children,
            List<Double> targets,
            double surplus,
            double availHPt,
            double gapPt,
            FitContext ctx
    ) {
        double remaining = surplus;
        List<Integer> absorbers = new ArrayList<>();
        for (int i = 0; i < children.size(); i++) {
            if (isAbsorbing(children.get(i), ctx)) {
                absorbers.add(i);
            }
        }
        if (!absorbers.isEmpty()) {
            return new Pair(children, distribute(targets, remaining, absorbers));
        }
        List<Integer> limited = new ArrayList<>();
        for (int i = 0; i < children.size(); i++) {
            if (isLimitedAbsorber(children.get(i), ctx)) {
                limited.add(i);
            }
        }
        if (!limited.isEmpty()) {
            List<Double> room = new ArrayList<>();
            for (int index : limited) {
                room.add(Math.max(0.0, targets.get(index) * LIMITED_ABSORB_MAX_RATIO - targets.get(index)));
            }
            double take = Math.min(remaining, room.stream().mapToDouble(Double::doubleValue).sum());
            if (take > 0) {
                targets = distributeCapped(targets, take, limited, room);
                remaining -= take;
            }
        }
        if (remaining <= 0) {
            return new Pair(children, targets);
        }
        double surplusAfterGap = remaining - gapPt;
        if (surplusAfterGap < availHPt * SURPLUS_MIN_RATIO) {
            List<Integer> body = bodyIndices(children);
            return new Pair(children, distribute(targets, remaining, body.isEmpty() ? allIndices(targets.size()) : body));
        }
        double spacerH = Math.min(surplusAfterGap, availHPt * MAX_SPACER_RATIO);
        double leftover = surplusAfterGap - spacerH;
        if (leftover > 0) {
            List<Integer> body = bodyIndices(children);
            targets = distribute(targets, leftover, body.isEmpty() ? allIndices(targets.size()) : body);
        }
        List<FlexNode> next = new ArrayList<>(children);
        next.add(fitSpacer("spacer-fit"));
        List<Double> nextTargets = new ArrayList<>(targets);
        nextTargets.add(spacerH);
        return new Pair(next, nextTargets);
    }

    private static Pair centerVertically(List<FlexNode> children, List<Double> targets, double surplus, double gapPt) {
        double budget = surplus - 2 * gapPt;
        if (budget <= 0) {
            return new Pair(children, targets);
        }
        double lead = budget * LEAD_SPACER_RATIO;
        List<FlexNode> next = new ArrayList<>();
        next.add(fitSpacer("spacer-fit-lead"));
        next.addAll(children);
        next.add(fitSpacer("spacer-fit-trail"));
        List<Double> nextTargets = new ArrayList<>();
        nextTargets.add(lead);
        nextTargets.addAll(targets);
        nextTargets.add(budget - lead);
        return new Pair(next, nextTargets);
    }

    private static FlexContainer fitSpacer(String nodeId) {
        return new FlexContainer("column", nodeId, List.of(), 0, null, null, 1.0);
    }

    private static List<Double> distribute(List<Double> targets, double amount, List<Integer> indices) {
        if (indices.isEmpty() || amount <= 0) {
            return new ArrayList<>(targets);
        }
        double weightSum = 0;
        for (int index : indices) {
            weightSum += targets.get(index);
        }
        List<Double> result = new ArrayList<>(targets);
        if (weightSum <= 0) {
            double share = amount / indices.size();
            for (int index : indices) {
                result.set(index, result.get(index) + share);
            }
            return result;
        }
        for (int index : indices) {
            result.set(index, result.get(index) + amount * targets.get(index) / weightSum);
        }
        return result;
    }

    private static List<Double> distributeCapped(List<Double> targets, double amount, List<Integer> indices, List<Double> rooms) {
        if (indices.isEmpty() || amount <= 0) {
            return new ArrayList<>(targets);
        }
        double roomSum = rooms.stream().mapToDouble(Double::doubleValue).sum();
        if (roomSum <= 0) {
            return new ArrayList<>(targets);
        }
        List<Double> result = new ArrayList<>(targets);
        for (int i = 0; i < indices.size(); i++) {
            int index = indices.get(i);
            result.set(index, result.get(index) + amount * rooms.get(i) / roomSum);
        }
        return result;
    }

    private static boolean isTitle(FlexNode node) {
        return node instanceof FlexLeaf leaf && leaf.textStyle() != null && FlexNormalize.TITLE_STYLES.contains(leaf.textStyle());
    }

    private static List<Integer> bodyIndices(List<FlexNode> children) {
        List<Integer> body = new ArrayList<>();
        for (int i = 0; i < children.size(); i++) {
            if (!isTitle(children.get(i))) {
                body.add(i);
            }
        }
        return body;
    }

    private static List<Integer> allIndices(int n) {
        List<Integer> indices = new ArrayList<>(n);
        for (int i = 0; i < n; i++) {
            indices.add(i);
        }
        return indices;
    }

    private static List<Double> targetsToGrows(List<Double> targets) {
        int n = targets.size();
        if (n == 0) {
            return List.of();
        }
        double total = targets.stream().mapToDouble(Double::doubleValue).sum();
        if (total <= 0) {
            return new ArrayList<>(java.util.Collections.nCopies(n, 1.0));
        }
        List<Double> grows = new ArrayList<>(n);
        for (double target : targets) {
            grows.add(Math.min(FlexNormalize.GROW_MAX, Math.max(FlexNormalize.GROW_MIN, target / total * n)));
        }
        return grows;
    }

    private static boolean isAbsorbing(FlexNode node, FitContext ctx) {
        if (node instanceof FlexLeaf leaf) {
            Map<String, Object> block = ctx.blockMap().get(leaf.blockId());
            return block != null && ABSORBING_TYPES.contains(Blocks.type(block));
        }
        if (node.isSpacer()) {
            return true;
        }
        FlexContainer container = (FlexContainer) node;
        for (FlexNode child : container.children()) {
            if (isAbsorbing(child, ctx)) {
                return true;
            }
        }
        return false;
    }

    private static boolean isLimitedAbsorber(FlexNode node, FitContext ctx) {
        if (node instanceof FlexLeaf leaf) {
            Map<String, Object> block = ctx.blockMap().get(leaf.blockId());
            return block != null && LIMITED_ABSORB_TYPES.contains(Blocks.type(block));
        }
        if (node.isSpacer()) {
            return false;
        }
        FlexContainer container = (FlexContainer) node;
        for (FlexNode child : container.children()) {
            if (isLimitedAbsorber(child, ctx)) {
                return true;
            }
        }
        return false;
    }

    private static double naturalHeightPt(FlexNode node, FitContext ctx) {
        if (node instanceof FlexLeaf leaf) {
            return leafHeightPt(leaf, ctx);
        }
        FlexContainer container = (FlexContainer) node;
        if (container.isSpacer() && container.children().isEmpty()) {
            return 0.0;
        }
        if (container.children().isEmpty()) {
            return MIN_LEAF_HEIGHT_PT;
        }
        List<Double> heights = new ArrayList<>();
        for (FlexNode child : container.children()) {
            heights.add(naturalHeightPt(child, ctx));
        }
        if (container.isRow()) {
            return heights.stream().mapToDouble(Double::doubleValue).max().orElse(0);
        }
        return heights.stream().mapToDouble(Double::doubleValue).sum()
                + Math.max(heights.size() - 1, 0) * container.gapPt();
    }

    @SuppressWarnings("unchecked")
    private static double leafHeightPt(FlexLeaf leaf, FitContext ctx) {
        Map<String, Object> block = ctx.blockMap().get(leaf.blockId());
        if (block == null) {
            return MIN_LEAF_HEIGHT_PT;
        }
        return switch (Blocks.type(block)) {
            case "text" -> measuredHeightPt(leaf, block, ctx, "text");
            case "bullets" -> measuredHeightPt(leaf, block, ctx, "bullets");
            case "table" -> {
                Object rows = block.get("rows");
                int n = rows instanceof List<?> list ? list.size() : 0;
                yield Math.max(MIN_LEAF_HEIGHT_PT, (1 + n) * TABLE_ROW_HEIGHT_PT);
            }
            case "cards" -> {
                Object items = block.get("items");
                double tallest = MIN_LEAF_HEIGHT_PT;
                if (items instanceof List<?> list) {
                    for (Object item : list) {
                        if (item instanceof Map<?, ?> map) {
                            String title = String.valueOf(map.get("title") == null ? "" : map.get("title"));
                            String desc = String.valueOf(map.get("desc") == null ? "" : map.get("desc"));
                            tallest = Math.max(
                                    tallest,
                                    CARD_BASE_HEIGHT_PT
                                            + Math.max(0, (title.length() + desc.length()) / 28) * CARD_LINE_HEIGHT_PT
                            );
                        }
                    }
                }
                yield tallest;
            }
            case "callout" -> PREFERRED_HEIGHT_PT.get("callout");
            default -> PREFERRED_HEIGHT_PT.getOrDefault(Blocks.type(block), MIN_LEAF_HEIGHT_PT);
        };
    }

    private static double measuredHeightPt(FlexLeaf leaf, Map<String, Object> block, FitContext ctx, String kind) {
        double widthPt = ctx.widthsPt().getOrDefault(leaf.blockId(), Geometry.SAFE_AREA_WIDTH_PT);
        BlockStyle boxStyle = Blocks.styleOf(block);
        BlockStyle.ResolvedBox box = BlockStyle.resolveBox(ctx.theme(), boxStyle);
        double[] avail = BlockStyle.contentRectPt(widthPt, Geometry.CANVAS_HEIGHT_PT, box.paddingPt());
        String defaultStyle = "bullets".equals(kind) ? "bullet" : "body";
        Theme.TextStyle style = BlockStyle.mergeTextStyle(
                ctx.theme(),
                leaf.textStyle() == null ? defaultStyle : leaf.textStyle(),
                boxStyle
        );
        double used = "bullets".equals(kind)
                ? TextMetrics.measureBullets(Blocks.items(block), style, avail[0], Geometry.CANVAS_HEIGHT_PT).heightPt()
                : TextMetrics.measureText(Blocks.text(block), style, avail[0], Geometry.CANVAS_HEIGHT_PT).heightPt();
        double chrome = 2 * TextMetrics.TEXTBOX_MARGIN_PT + 2 * box.paddingPt();
        return Math.max(MIN_LEAF_HEIGHT_PT, used * BREATHING + chrome);
    }
}
